tuxemon/platform/events.py

```diff
@@ -19,59 +19,118 @@
 # You should have received a copy of the GNU General Public License
 # along with this program. If not, see <http://www.gnu.org/licenses/>.
 #
+from __future__ import annotations
+from typing import (
+    Any,
+    Generator,
+    Generic,
+    Mapping,
+    Optional,
+    Sequence,
+    TypeVar,
+)
+from abc import ABC, abstractmethod
 
 
-class EventQueueHandler:
-    """Event QueueHandler for different platforms
+_InputEventType = TypeVar("_InputEventType", contravariant=True)
+
+
+class EventQueueHandler(ABC):
+    """Event QueueHandler for different platforms.
 
     * Only one per game
     * Sole manager of platform events of type
     """
 
-    def release_controls(self):
-        """Send virtual input events which release held buttons/axis
+    _inputs: Mapping[int, Sequence[InputHandler[Any]]]
+
+    def release_controls(self) -> Generator[PlayerInput, None, None]:
+        """
+        Send virtual input events which release held buttons/axis.
+
+        After this frame, held/triggered inputs will return to previous state.
 
-        After this frame, held/triggered inputs will return to previous state
+        Yields:
+            Inputs to release all buttons.
 
-        :return:
         """
-        for key, value in self._inputs.items():
+        for value in self._inputs.values():
             for inp in value:
                 yield from inp.virtual_stop_events()
 
-    def process_events(self):
+    @abstractmethod
+    def process_events(self) -> Generator[PlayerInput, None, None]:
+        """
+        Process all pygame events.
+
+        * Should never return pygame-unique events
+        * All events returned should be Tuxemon game specific
+        * This must be the only function to get events from the pygame event
+          queue
+
+        Yields:
+            Game events.
+
+        """
         raise NotImplementedError
 
 
-class InputHandler:
-    """Enables basic input device with discrete inputs"""
+class InputHandler(ABC, Generic[_InputEventType]):
+    """
+    Enables basic input device with discrete inputs.
+
+    Parameters:
+        event_map: Mapping of original identifiers to button identifiers.
+
+    """
 
-    event_type = None
-    default_input_map = None
+    default_input_map: Mapping[Optional[int], int]
 
-    def __init__(self, event_map=None):
+    def __init__(
+        self,
+        event_map: Optional[Mapping[Optional[int], int]] = None,
+    ) -> None:
         if event_map is None:
-            event_map = self.default_input_map.copy()
```

Comment:
this copy should be kept.  if the event map is changed, then the default would be changed for any other devices that are using it.  the config file can reconfigure controls, so this means it won't be possible to have different controls for other players....think split screen.

Reply:
The event map is not changed anywhere. It is only used in the files modified in this PR.

Also, I have used an immutable type, so Mypy should flag as an error any attempt to change it.

Reply:
fair enough